hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,420 @@
+//!
+//! # Releasing the DMAC
+//!
+//! Using the [`DmaController::release`](DmaController::release) method will
+//! deinitialize the DMAC and return the underlying PAC object.
+
+#[cfg(any(feature = "samd11", feature = "samd21"))]
+pub use crate::target_device::dmac::chctrlb::{
+    LVL_A as PriorityLevel, TRIGACT_A as TriggerAction, TRIGSRC_A as TriggerSource,
+};
+
+#[cfg(any(
+    feature = "samd51",
+    feature = "same51",
+    feature = "same53",
+    feature = "same54"
+))]
+pub use crate::target_device::dmac::channel::{
+    chctrla::{
+        BURSTLEN_A as BurstLength, THRESHOLD_A as FifoThreshold, TRIGACT_A as TriggerAction,
+        TRIGSRC_A as TriggerSource,
+    },
+    chprilvl::PRILVL_A as PriorityLevel,
+};
+
+use super::{
+    channel::{new_chan, Channel, Uninitialized},
+    DESCRIPTOR_SECTION, WRITEBACK,
+};
+use crate::target_device::{DMAC, PM};
+
+/// Initialized DMA Controller
+pub struct DmaController {
+    dmac: DMAC,
+}
+
+impl AsMut<DMAC> for DmaController {
+    /// Return a mutable reference to the underlying DMAC object exposed by the
+    /// PAC.
+    fn as_mut(&mut self) -> &mut DMAC {
+        &mut self.dmac
+    }
+}
+
+impl AsRef<DMAC> for DmaController {
+    /// Return an immutable reference to the underlying DMAC object exposed by
+    /// the PAC.
+    fn as_ref(&self) -> &DMAC {
+        &self.dmac
+    }
+}
+
+impl DmaController {
+    /// Initialize the DMAC and return a DmaController object useable by
+    /// [`DmaTransfer`](super::transfer::DmaTransfer)'s. By default, all
+    /// priority levels are enabled unless subsequently disabled using the
+    /// `level_x_enabled` methods.
+    pub fn init(mut dmac: DMAC, _pm: &mut PM) -> Self {
+        // ----- Initialize clocking ----- //
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        {
+            // Enable clocking
+            _pm.ahbmask.modify(|_, w| w.dmac_().set_bit());
+            _pm.apbbmask.modify(|_, w| w.dmac_().set_bit());
+        }
+
+        Self::swreset(&mut dmac);
+
+        // SAFETY this is safe because we write a whole u32 to 32-bit registers,
+        // and the descriptor array addesses will never change since they are static.
+        // We just need to ensure the writeback and descriptor_section addresses
+        // are valid.
+        unsafe {
+            dmac.baseaddr
+                .write(|w| w.baseaddr().bits(DESCRIPTOR_SECTION.as_ptr() as u32));
+            dmac.wrbaddr
+                .write(|w| w.wrbaddr().bits(WRITEBACK.as_ptr() as u32));
+        }
+
+        // ----- Select priority levels ----- //
+        // TODO selectively enable priority levels
+        // right now we blindly enable all priority levels
+        dmac.ctrl.modify(|_, w| {
+            w.lvlen3()
+                .set_bit()
+                .lvlen2()
+                .set_bit()
+                .lvlen1()
+                .set_bit()
+                .lvlen0()
+                .set_bit()
```

Comment:
This is another example of `rustfmt` murdering the code. I would separate these into individual field changes.